---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
rm(list = ls())

library(ggplot2)
library(plyr)

strReverse <- function(x) sapply(lapply(strsplit(x, NULL), rev), paste, collapse="")

today = Sys.Date()
```

```{r}
# load pairs of k-mers
kmer_pairs = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/poolseq-kmers/kmerpairs_sequences.tsv", col.names = c("k1"))

kmer_covs = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/poolseq-kmers/kmerpairs_coverages.tsv", col.names = c("c1", "c2"))

kmer_pairs = cbind(kmer_pairs, kmer_covs)

# choose kmer count cutoff
all_counts = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/poolseq-kmers/kmer_counts.txt", col.names = c("k1", "c1"))

ggplot(all_counts[(all_counts$c1 > 5),], aes(x = c1)) +
  geom_histogram(binwidth = 1) +
  theme_classic()

# filter kmer pairs by coverage
#kmer_pairs = kmer_pairs[(kmer_pairs$c1 >=5 & kmer_pairs$c2 >= 5),]

#kmer_pairs = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/kmers_poolseq/kmer_pairs.txt", fill = T, col.names = c("pos", "k1", "c1", "k2", "c2", "k3", "c3", "k4", "c4"))

# exclude sites with non-isolated snps
#kmer_pairs = kmer_pairs[!(kmer_pairs$k1 == ""),]
#kmer_pairs = kmer_pairs[!(kmer_pairs$k2 == ""),]
#kmer_pairs = kmer_pairs[(kmer_pairs$k3 == ""),]

#kmer_pairs = kmer_pairs[(mapply(function(x,y) #sum(x!=y),strsplit(kmer_pairs$k1,""),strsplit(kmer_pairs$k2,"")) == 1),]

#kmer_pairs = kmer_pairs[(mapply(function(x,y) which(x!=y),strsplit(kmer_pairs$k1,""),strsplit(kmer_pairs$k2,"")) == 16),]
       
#kmer_pairs = kmer_pairs[(c(1000, diff(kmer_pairs$pos)) > 15),] # remove non-isolated snps
           
#kmer_pairs$c1 = as.numeric(kmer_pairs$c1)
#kmer_pairs$c2 = as.numeric(kmer_pairs$c2)

# calculate minor allele frequency
kmer_pairs$kmer_maf = pmin(kmer_pairs$c1, kmer_pairs$c2)/(kmer_pairs$c1 + kmer_pairs$c2)

ggplot(kmer_pairs, aes(x = kmer_maf)) +
  geom_density() +
  theme_classic()

# load in actual allele freqs from simulations
true_af = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/poolseq-kmers/slim_allele_freqs.txt", col.names = c("chrom", "pos", "af", "ac"))

true_af$ac = as.numeric(true_af$ac)

true_af = true_af[complete.cases(true_af),]

# allele frequency spectrum
ggplot(true_af, aes(x = ac)) +
  geom_histogram(binwidth = 1) +
  theme_classic()

# calculate minor allele frequencies
true_af$af = as.numeric(true_af$af)
true_af$snp_maf = pmin(true_af$af, 1 - true_af$af)

# get key for position to k-mers
pos_kmer_key = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/poolseq-kmers/center_kmer_pairs.txt", fill = T, col.names = c("pos", "k1", "k2", "k3", "k4"))

#pos_kmer_key = pos_kmer_key[(pos_kmer_key$V4 == ""),]
#pos_kmer_key = pos_kmer_key[(pos_kmer_key$V2 != ""),]
#pos_kmer_key = pos_kmer_key[(pos_kmer_key$k3 == ""),]

pos_kmer_key$leftk = substr(pos_kmer_key$k1, 1, 15)
pos_kmer_key$rightk = substr(pos_kmer_key$k1, 17, 31)

# compare snp and k-mer mafs
kmer_pairs$leftk = substr(kmer_pairs$k1, 1, 15)
kmer_pairs$rightk = substr(kmer_pairs$k1, 17, 31)

compare_maf = merge(pos_kmer_key, kmer_pairs, by = c("leftk", "rightk"))

# kmer_pairs$leftk = chartr("ATGC","TACG",kmer_pairs$leftk)
# kmer_pairs$rightk = chartr("ATGC","TACG",kmer_pairs$rightk)
# 
# kmer_pairs$leftk = strReverse(kmer_pairs$leftk)
# kmer_pairs$rightk = strReverse(kmer_pairs$rightk)

# get snps that pair with reverse complements
pos_kmer_key$leftk_rc = chartr("ATGC","TACG",pos_kmer_key$leftk)
pos_kmer_key$rightk_rc = chartr("ATGC","TACG",pos_kmer_key$rightk)

pos_kmer_key$leftk_rc = strReverse(pos_kmer_key$leftk_rc)
pos_kmer_key$rightk_rc = strReverse(pos_kmer_key$rightk_rc)

pos_kmer_key$rightk = pos_kmer_key$leftk_rc 
pos_kmer_key$leftk = pos_kmer_key$rightk_rc

compare_maf_rc = merge(pos_kmer_key, kmer_pairs, by = c("leftk", "rightk"))

# merge reverse complement and regular sets
any(duplicated(c(compare_maf$pos, compare_maf_rc$pos)))

compare_maf = rbind(compare_maf, compare_maf_rc[,c(-8,-9)])

compare_maf = merge(true_af[,c("pos", "snp_maf")], compare_maf, by = "pos")
#compare_maf = compare_maf[!is.na(compare_maf$snp_maf),]

# include varscan calls
varscan = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/poolseq-kmers/calls.tsv", header = T)

varscan$snp_maf_varscan = pmin(varscan$Reads1, varscan$Reads2)/(varscan$Reads1 + varscan$Reads2)

names(varscan)[2] = "pos"

# how many
calc_rates = function(truth, test, sample_space){
  negatives = sample_space[!(sample_space %in% truth)]
  positives = truth
  true_positives = sum(test %in% positives)
  false_positives = sum(!(test %in% positives))
  true_negatives = sum(!(negatives %in% test))
  false_negatives = sum(negatives %in% test)
  
  # accuracy
  acc = (true_positives + true_negatives)/(true_positives + true_negatives + false_positives + false_negatives)
  
  # true positives rate (sensitivity)
  tpr = true_positives/length(positives)

  # false positive rate (false alarm)
  fpr = false_positives/length(positives)

  # true negative rate (specificity)
  tnr = true_negatives/length(negatives)

  # false negative rate (miss)
  fnr = 1 - true_negatives/length(negatives)

  return(c(acc, tpr, fpr, tnr, fnr))
}

calc_rates(true_af$pos, varscan$pos, 1:1e6)

calc_rates(true_af$pos, compare_maf$pos, 1:1e6)

# add in varscan calls
compare_maf = merge(compare_maf, varscan, by = "pos")

# do some filtering
# remove low coverage k-mer pairs
# remove tri-allelic k-mers
compare_maf = compare_maf[(compare_maf$c1 >=5 & compare_maf$c2 >= 5),]
compare_maf = compare_maf[(compare_maf$k3 == ""),]

# round k-mer mafs to nearest increment 
cor.test(compare_maf$snp_maf, compare_maf$kmer_maf, method = "spearman")

cor.test(compare_maf$snp_maf, compare_maf$snp_maf_varscan, method = "spearman")

ggplot(compare_maf, aes(x = snp_maf, y = kmer_maf)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0) +
  theme_classic() +
  labs(x = "True MAF", y = "Predicted MAF from k-mers")

ggplot(compare_maf, aes(x = snp_maf, y = snp_maf_varscan)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0) +
  theme_classic() +
  labs(x = "True MAF", y = "Predicted MAF from VarScan")

# what about for k-mers above some frequency threshold
kmer_test = cor.test(compare_maf$snp_maf, compare_maf$kmer_maf, method = "spearman")

ggplot(compare_maf, aes(x = snp_maf, y = kmer_maf)) +
  geom_density_2d_filled() +
  geom_abline(slope = 1, intercept = 0) +
  theme_classic() +
  theme(text = element_text(size = 20)) +
  labs(x = "True MAF", y = "Predicted MAF from k-mers", title = paste("rho =", signif(kmer_test$estimate, 4), "; p =", paste(format(signif(kmer_test$p.value, 2), digits = 2))))

ggsave(paste("/mnt/home/robe1195/Josephs_Lab_Projects/poolseq-kmers/results/", today,  "/","kmer_maf_vs_truth_pool-seq.png", sep = ""), height= 7, width = 8)

varscan_test = cor.test(compare_maf$snp_maf, compare_maf$snp_maf_varscan, method = "spearman")

ggplot(compare_maf, aes(x = snp_maf, y = snp_maf_varscan)) +
  geom_density_2d_filled() +
  geom_abline(slope = 1, intercept = 0) +
  theme_classic() +
  theme(text = element_text(size = 20)) +
  labs(x = "True MAF", y = "Predicted MAF from VarScan", title = paste("rho =", signif(varscan_test$estimate, 4), "; p =", paste(format(signif(varscan_test$p.value, 2), digits = 2))))

ggsave(paste("/mnt/home/robe1195/Josephs_Lab_Projects/poolseq-kmers/results/", today,  "/","varscan_maf_vs_truth_pool-seq.png", sep = ""), height= 7, width = 8)

# smudgeplot
ggplot(compare_maf, aes(x = kmer_maf, y = (c1+c2)/81.670932)) +
  geom_density_2d_filled() +
  #geom_hline(yintercept = 50) +
  theme_classic() +
  theme(text = element_text(size = 20)) +
  labs(x = "Minor k-mer frequency", y = "Total pair coverage")

ggsave(paste("/mnt/home/robe1195/Josephs_Lab_Projects/poolseq-kmers/results/", today,  "/","smudgeplot.png", sep = ""), height= 7, width = 8)

# site-frequency spectrum
plotdata = melt(as.data.table(compare_maf[,c("pos", "snp_maf", "kmer_maf", "snp_maf_varscan")]), id = "pos")

ggplot(plotdata, aes(x = value, fill = variable)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(paste("/mnt/home/robe1195/Josephs_Lab_Projects/poolseq-kmers/results/", today,  "/","kmer_vs_snp_sfs.png", sep = ""), height= 7, width = 8)

# folded sfs
kmer_pairs = kmer_pairs[(kmer_pairs$c1 >=5 & kmer_pairs$c2 >= 5),]

folded_sfs = data.frame(
    category=factor(c(rep("truth", times = nrow(true_af)), rep("kmer", times = nrow(kmer_pairs)), rep("varscan", times = nrow(varscan)))),
    value=c(true_af$snp_maf,kmer_pairs$kmer_maf, varscan$snp_maf_varscan)
)

ggplot(folded_sfs, aes(x = value, fill = category)) +
  geom_density(alpha = 0.5) +
  theme_classic() +
  theme(text = element_text(size = 20)) +
  labs(x = "minor allele frequency")

ggsave(paste("/mnt/home/robe1195/Josephs_Lab_Projects/poolseq-kmers/results/", today,  "/","kmer_vs_snp_sfs.png", sep = ""), height= 7, width = 8)

ggplot(true_af, aes(x = snp_maf)) +
  geom_density() +
  theme_classic()

ggplot(kmer_pairs, aes(x = kmer_maf)) +
  geom_density() +
  theme_classic()

ggplot(kmer_pairs, aes(x = kmer_maf)) +
  geom_density() +
  theme_classic()

ggplot(true_af[(true_af$snp_maf >= 0.05),], aes(x = snp_maf)) +
  geom_density() +
  theme_classic()

ggplot(compare_maf, aes(x = kmer_maf)) +
  geom_histogram(binwidth = 0.05) +
  theme_classic()

ggplot(compare_maf, aes(x = snp_maf)) +
  geom_histogram(breaks = seq(from = 0, to = 0.5, by = 0.02)) +
  theme_classic()

ggplot(compare_maf, aes(x = kmer_maf)) +
  geom_histogram(breaks = seq(from = 0, to = 0.5, by = 0.02)) +
  theme_classic()

# calculate error
compare_maf$error = abs(compare_maf$kmer_maf - compare_maf$snp_maf)

summary(compare_maf$error)

sum(compare_maf$error < 0.1)/nrow(compare_maf)

ggplot(confident_sites, aes(x = total_cov, y = error)) +
  geom_point() +
  theme_classic()

#
compare_maf$snp_maf_bins = cut(compare_maf$snp_maf, breaks = seq(from = 0, to = 0.5, by = 0.05))
compare_maf$kmer_maf_bins = cut(compare_maf$kmer_maf, breaks = seq(from = 0, to = 0.5, by = 0.05))

observed <- table(compare_maf$kmer_maf_bins)  # Observed frequencies
expected <- table(compare_maf$snp_maf_bins)  # Expected frequencies

observed = observed[(expected != 0)]
expected = expected[(expected != 0)]

# Calculate Chi-Square statistic manually
chi_sq_statistic <- sum((observed - expected)^2 / expected)
df <- length(observed) - 1
p_value <- 1 - pchisq(chi_sq_statistic, df)

# Print results
print(paste("Chi-Square Statistic:", chi_sq_statistic))
print(paste("Degrees of Freedom:", df))
print(paste("P-value:", p_value))

```

```{r}
kmer_counts = read.table("/mnt/scratch/robe1195/Josephs_Lab_Projects/kmers_poolseq/kmer_counts.txt", col.names = c("k1", "c1"))

ggplot(kmer_counts, aes(x = c1)) +
  geom_histogram(binwidth = 1) +
  geom_vline(xintercept = 5) +
  theme_classic()

# loop over k-mers to find ones that putative snps
alphabet = c("A","T","G","C")

results = list()

j = 1
while(nrow(kmer_counts) > 0){
  # get focal k-mer
  focal_kmer = kmer_counts[1,"k1"]
  focal_counts = kmer_counts[1,"c1"]
  print(focal_kmer)

  # generate variant k-mers
  central_base = substr(focal_kmer, 16, 16)

  var_bases = alphabet[!(alphabet %in% central_base)]

  variant_kmers = rep(focal_kmer, times = 3)

  for(i in 1:3){
    var_base = var_bases[i]
    substr(variant_kmers[i], 16, 16) = var_base
  }

  # add reverse complements
  variant_kmers = c(variant_kmers, chartr("ATGC","TACG",variant_kmers))

  # 
  variant_kmers_index = which(kmer_counts$k1 %in% variant_kmers)
  print(variant_kmers_index)

  # store 
  found_variant_kmers = kmer_counts[variant_kmers_index,"k1"]
  found_variant_counts = kmer_counts[variant_kmers_index,"c1"]
  
  if(!identical(found_variant_kmers, character(0))){
      results[[j]] = data.frame(focal_kmer,
                            focal_counts,
                            found_kmers = paste(found_variant_kmers, sep = ";"),
                            found_counts = paste(found_variant_counts, sep = ";"))

    j = j + 1
  }

  # remove focal and variant k-mers from count table
  kmer_counts = kmer_counts[c(-1, -1*variant_kmers_index),]

}


```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
